pulp_ansible/app/tasks/signature.py

```diff
@@ -0,0 +1,114 @@
+            ContentSaver(),
+        ]
+        return pipeline
+
+
+class CollectionSigningFirstStage(Stage):
+    """
+    This stage signs the content and creates CollectionVersionSignatures.
+    """
+
+    def __init__(self, content, signing_service):
+        """Initialize Signing first stage."""
+        super().__init__()
+        self.content = content
+        self.signing_service = signing_service
+        # Maybe make this value a pulp_ansible constant setting?
+        self.semaphore = asyncio.Semaphore(10)
+
+    async def sign_collection_version(self, collection_version):
+        """Signs the collection version."""
+        # Limits the number of subprocesses spawned/running at one time
+        # Should everything be under the semaphore or just the signing part?
+        async with self.semaphore:
+            # We use the manifest to create the signature
+            with tempfile.TemporaryDirectory() as d:
+                # OpenPGP doesn't take filename into account for signatures, not sure about others
+                with open(f"{d}/MANIFEST.json", "w") as t:
+                    json.dump(collection_version.manifest, t)
+                result = await self.signing_service.asign(t.name)
+                # Is it always a guarantee that the result will have a signature where the file is?
+                with open(result["signature"], "rb") as sig:
+                    data = sig.read()
+            cv_signature = CollectionVersionSignature(
+                data=data,
+                digest=hashlib.sha256(data),
+                signed_collection=collection_version,
+                pubkey_fingerprint=self.signing_service.pubkey_fingerprint,
+                signing_service=self.signing_service,
+            )
+            dc = DeclarativeContent(content=cv_signature)
+            await self.progress_report.aincrement()
+            await self.put(dc)
+
+    async def run(self):
+        """Signs collections if they have not been signed with key."""
+        tasks = []
+        # Filter out any content that already has a signature with pubkey_fingerprint
+        current_signatures = CollectionVersionSignature.objects.filter(
+            pubkey_fingerprint=self.signing_service.pubkey_fingerprint
+        )
+        new_content = self.content.exclude(signatures__in=current_signatures)
+        ntotal = await sync_to_async(new_content.count)()
+        nmsg = _("Signing new CollectionVersions")
+        async with ProgressReport(message=nmsg, code="sign.new.signature", total=ntotal) as p:
+            self.progress_report = p
+            async for collection_version in sync_to_async_iterable(new_content.iterator()):
+                tasks.append(asyncio.create_task(self.sign_collection_version(collection_version)))
+            await asyncio.gather(*tasks)
+
+        # Add any signatures already present in Pulp if part of content list
+        present_content = current_signatures.filter(signed_collection__in=self.content)
```

Comment:
This is lying a little bit. If the present signature is already in the repo version, it will still be counted here, but no actual content will be added to the new repo version.

Reply:
You're right. I'll try a different query which excludes signatures already in the repo_version.